### DISCUSSION TITLE
far-field extraction in cylindrical coordinates - Different approaches

### DISCUSSION THREAD

**Discussion by kalosu** [NanoComp/meep#2842] [Q&A]
Hello there meep community, 

I have a question related to the two methods available to extract far-field information from a near-field object.

I am aware of ```get_farfields``` and ```get_farfield```, in where for the first one I can get the far-field over an extended region, whereas for the second approach I get the far-field for an specific ```x``` coordinate. 

Is there any significant and important differences between these two approach (apart from the extended region vs pointwise evaluations) that I should consider? (My evaluations are done in cylindrical coordinates)

I have a dipole embedded in a planarized structure. I use these two far-field extraction approaches to obtain the far-field at a given distance.

Using ```get_farfields``` I obtain the far-field over a plane with constant z value, defined in the rz plane.
Using ```get_farfield``` I obtain the far-field over a section of a circle defined over the rz plane.

After extracting the far-fields, I transform the components into cartesian ones and finally, I try to estimate the wavefront deviations with respect to a reference spherical wavefront.

Basically what I want to do is to evaluate the difference of the obtained wavefront in the far-field with respect to an ideal spherical wavefront.

The issue now is that I obtain different results depending on which far-field extraction approach I use.

With the ```get_farfields``` method I get the following optical path difference plot:
![meep_dip_ips_pva_gold_silica_opd_eval_ff](https://github.com/NanoComp/meep/assets/49510503/62b4b17c-e5a6-4818-b21f-86942b0f9ff7)

Whereas with the ```get_farfield``` method I get the following results:
![meep_dip_pva_ips_gold_silica_ff_circ_phase_diff_opd](https://github.com/NanoComp/meep/assets/49510503/ef238f06-4916-4637-8999-c189bab91e6c)

There is around one order of magnitude difference between these two configurations.
Ideally, I would expect to have the same deviations for both approaches (the reference wavefront is an ideal spherical wavefront in both cases. This tells me that the deviations come from the far-field results)

Any idea about what could be happening here?

**Comment by kalosu** [comment]
Maybe one more question regarding above situation.

When I perform a near-to-far-field evaluation, what is the reference origin point that is used? 

Is it the upper plane of my near-field monitor, is it the 0 of my simulation domain, is the location of the point source?

**Reply by stevengj** [reply]
I don't know what you mean by "reference" point.  Every near-field point is treated as a source that is propagated to the far-field point(s) and summed.

The coordinates of the far-field point are given relative to the (0,0,0) origin of the simulation domain, if that's what you mean.

**Comment by stevengj** [comment]
`get_farfields` and `get_farfield` should compute the same thing given the same points.  The only difference should be that `get_farfields` is more efficient if you need many points.

It sounds like you are computing the fields at different points in the two methods.  Can you try computing the *same* point using both routines and verify that they give the same result?

**Comment by kalosu** [comment]
Hello @stevengj,

Yes, I was talking about the reference point for the far-field evaluation coordinates. If I understand it correctly, when I try to evaluate the far-field at r=0, z=100, then, the coordinate (0,0,100) will be estimated relative to the simulation domain at (0,0,0) right? And not relative to the near-field monitor right? 

I know that I can shift the origin point for my simulation domain using the ```geometry_center``` option while creating the ```Simulation``` object. If I wanted to compare the obtained wavefront in the far-field to a reference spherical wavefront, what would you suggest to use as the (0,0,0) point of my simulation domain?

My domain looks something like this:
![cell_NV_pva_ips_gold_silica](https://github.com/NanoComp/meep/assets/49510503/f582badc-2b15-49b3-aeb4-9b87d1eb2056)

The layers that I have are basically (from top to bottom):
PML, air, PVA, gold, silica, PML

Also, my dipole source is embedded within the PVA layer, 50 nm, above the PVA-gold interface.

I would assume that if I want to estimate the wavefront "aberrations" at the far-field in air (basically the deviations of the obtained far-field wavefront with respect to a reference spherical wavefront), I would then need to use the air/PVA interface as the origin of my coordinate system so that the far-field points are referrenced to the air-PVA interface right? 

Thinking in terms of the Hyugens principle, I should have a spherical wavefront originating from the air-PVA interface, isnt it?

Or would it make more sense to set the reference at the position of my dipole emitter? At the end, the light is in fact being originated from the dipole, i.e, here is where the spherical wavefront would start to radiate from no?

Maybe you have some comments on this.

**Comment by oskooi** [comment]
If you are computing your far fields at a sufficiently large distance from the simulation domain (i.e., hundreds of optical wavelengths) then the simulation domain can be treated essentially as a **point-source emitter**. Note that in the [tutorials involving the radiation pattern in cylindrical coordinates](https://meep.readthedocs.io/en/latest/Python_Tutorials/Near_to_Far_Field_Spectra/#extraction-efficiency-of-a-collection-of-dipoles-in-a-disc), the far fields are computed along a quarter circle with an arbitrary radius of `FARFIELD_RADIUS_UM = 1e6 * WAVELENGTH_UM`. #2463 describes an approach to simplify this type of calculation.

**Reply by kalosu** [reply]
Hi @oskooi,

Thanks a lot for your feedback. I have given it a try by defining a sufficiently long distance from the simulation domain (300 microns for a wavelength of 0.640 nm). 

If I extract the phase over a planar region (constant z), I have no problem in the results that I obtain. In fact, I obtain the same results regarless of whether I use ```get_farfields``` or ```get_farfield```. 

However, once I try to extract the far-fields using a quarter of a circle, there are some issues with the computed phase distributions.

After constructing $E_x$ from $E_r$ and $E_p$, I am able to obtain the total $E_x$ spanned over the entire upper half sphere surface (a half sphere with a radius equal to 300). However,I observe some problems from evaluating the wrapped phase directly from the complex amplitude term $E_x$.

For a perfect spherical wave, I should expect that all phase values are the same, since by definition, the phase distribution for an ideal spherical wavefront is the same all over the spherical surface (i.e, in this case the surface used for evaluating the far-field)

For the structure that I have, I get the following phase wavefront profiles (wrapped phase):

![ex_phase](https://github.com/NanoComp/meep/assets/49510503/8c28aef4-3ee3-4200-a31f-cbf903fa0415)

You can observe how there is a strong phase jump from the center towards the outer part of the evaluation region(the axes are the azimuthal and polar angles used for extracting the far-field information)

While using the planar surface evaluation region (constant z), I do not observe these strong phase discontinuties and the obtained wavefront is pretty close to an ideal spherical wavefront (some minor deviations due to the refraction at the interfaces defining my structure).

What I want to do at the end, is to have an estimate on the wavefront "aberrations". However, this strong phase jump in the spherical surface case does not allow me to properly estimate the deviations with respect to the "un aberrated" ideal spherical wavefront.

Any ideas about what I could do in this case?

**Reply by oskooi** [reply]
> After constructing $E_x$ from $E_r$ and and $E_p$, I am able to obtain the total $E_x$ spanned over the entire upper half sphere surface (a half sphere with a radius equal to 300). However,I observe some problems from evaluating the wrapped phase directly from the complex amplitude term $E_x$.

What does the phase map over the hemisphere look like for the *radially* outward fields (this is not the same thing as $E_r$)? Perhaps you can share the script you used to generate these phasemaps.

**Comment by kalosu** [comment]
Hi again friends from Meep,

I am wondering if you have any comments about what could be happening with the wrapped phase profile shown above in my last post?

As I mentioned before, my goal is to be able to evaluate the deviations of the field wavefront with respect to a reference spherical wavefront.

In the field of classical imaging systems design, these deviations are known as aberrations. This information is really useful because it really tells you about the direction at which the light is essentially propagating.

Now for electromagnetic fields as in the case of the field components extracted from meep, it is not so simple to have an estimate on these "aberrations". The only thing that I know, is that the vector E and H fields are both perpendicular to each other and perpendicular to the wave vector k  at every point in space. (At least for transversal field distribution or in free space propagation regions. This does not hold for guided modes in a cavity). In a similar manner, the Poynting vector indicates the direction of energy flow and for the case of linear, isotropic mediums, both k and the Poynting vector are oriented along the same direction.

Knowing this, I tried to estimate the deviations of the k vector directions (via the Poynting vector) with respect to the normals of a spherical surface. If both vectors (the unit vector from the poynting vector and the normal vectors to the spherical surface) are pointing along the same direction, this would serve as a good indication that my wavefront is spherical in shape right?

This is somehow what can be observed in the following figure:


<img width="1316" alt="normals_to_sphere_dot_product_pointing_vector" src="https://github.com/NanoComp/meep/assets/49510503/473d0fed-ae83-4cd8-b7d9-176c84d00fd8">

In this case, you can observe (top left plot) that there is a really good overlap of the direction vectors from the Poynting vectors and the normals to the spherical surface. (just to clarify, the field components are also extracted over a spherical surface. But if at these surface points, the energy flow, given by the direction of the Poyinting vector, was not along the normals of the spherical surface, then I would have a reduced overlap between both unit vectors right?)

However, I still cannot see how the wrapped phase of the Ex component as shown in my previous post has that sharp jump at the center. Once I try to unwrap this phase profile and again try to estimate the deviations with respect to an ideal spherical phase term, the deviations are somehow larger (with a maximum deviation of around 0.8 lambdas) 

Any idea on how could I have a better way to estimate the wavefront deviations?
